arm64: cache: refine arch_dcache_range()

JiafeiPan opened the pull request:
Move all assembly code to c code.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
arm64: cache: refine arch_dcache_range()

Moved all assembly code to c code. Fixed arch_dcache_line_size_get()
to get dcache line size by using "4 << dminline" and don't consider
CWG according to sample code in cotexta-v8 programer guider.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

carlocaione commented on arch/arm64/core/cache.c:
Why not directly using `arch_dcache_line_size_get()`?

JiafeiPan replied:
I am not sure why use the following code to get line size in arch_dcache_line_size_get()?

dcache_line_size = cwg ? 4 << cwg : 4 << dminline;

I find it use 4 << DminLine to get the mininual line size in  sample code of "Cleaning by Virtual Address" in cortex a-v8 programer guider.

JiafeiPan replied:
@MrVan , could you please address why use cwg field to get dcache line size, I found in the other SW component such as tf-a and uboot, and also in ARM's sample code, it only get the minimal dcache line size by using dminline ? thanks.

carlocaione replied:
Indeed not clear what the CWG is for there.

JiafeiPan replied:
Yes, I removed CSW and use dminline only.

JiafeiPan replied:
Typo, removed CWG.

npitre commented:
There was a `dsb` at the end of the loop in the assembly version.
Don't you need it in the C version as well?

JiafeiPan commented:
> There was a `dsb` at the end of the loop in the assembly version.
> Don't you need it in the C version as well?

Yes,  we need dsb there, updated.

JiafeiPan committed:
arm64: cache: fix arch_dcache_all()

Add data barrier before and after dcachle flush or clean,
and restore to data cache level 0 after all ops.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
